fdbserver/VFSAsync.h

```diff
@@ -0,0 +1,92 @@
+ */
+
+#include "sqlite/sqlite3.h"
+#include <string>
+#include <map>
+#include "fdbrpc/IAsyncFile.h"
+
+/*
+** When using this VFS, the sqlite3_file* handles that SQLite uses are
+** actually pointers to instances of type VFSAsyncFile.
+*/
+typedef struct VFSAsyncFile VFSAsyncFile;
+struct VFSAsyncFile {
+	sqlite3_file base;              /* Base class. Must be first. */
+	int flags;
+	std::string filename;
+	Reference<IAsyncFile> file;
+
+    // The functions setInjectedError() and checkInjectedError() use an INetwork global to store the last
+    // return code from VFSAsyncFile method resulting from catching an injected Error exception.  This allows
+    // callers of the SQLite API to determine if an error code returned appears to be due to an injected
+    // error in simulation.
+    //
+    // This scheme is not perfect, as it is possible for non-injected errors to occur after injected errors
+    // and be incorrectly recognized as injected.  This problem already existed, however, as the previous scheme
+    // assumed that any SQLite error that occurred after any injected error in the simulated process was
+    // itself injected.
+    //
+    // Unfortunately, there is no easy or reliable way to plumb the injectedness of an error though the return
+    // path of VFSAsyncFile -> SQLite -> SQLite API calls made by KeyValueStoreSQLite.
+    //
+    // An attempt was made to store injected errors in VFSAsyncFile instances and expose a SQLiteDB's file
+    // instances via the SQLite API.  This failed, however, because sometimes files are opened, encounter an
+    // error, and are closed within the lifetime of one SQLite API call so the caller never has an opportunity
+    // to access the VFSAsyncFile object or its injected error state.
+    //
+    // An attempt was also made to match SQLite API return codes to VFSAsyncFile injected error return codes on 
+    // a 1:1 basis, only flagging a code as rejected if it matches the last injected error code and only once.
+    // This would have been more accurate (though coincidences could occur).  This scheme also failed, however,
+    // because sometimes errors from SQLite APIs are temporarily ignored, relying on a subsequent API call to error,
+    // however this error could be different and would not have been produced directly by VFSAsyncFile.
+    //
+    static void setInjectedError(int64_t rc) {
+        g_network->setGlobal(INetwork::enSQLiteInjectedError, (flowGlobalType)rc);
+        TraceEvent("VFSSetInjectedError").detail("ErrorCode", rc).detail("NetworkPtr", (void *)g_network).backtrace();
+    }
+
+    static bool checkInjectedError() {
+        // Error code is only checked for non-zero because the SQLite API error code after an injected error
+        // may not match the error code returned by VFSAsyncFile when the inject error occurred.
+        bool e = g_network->global(INetwork::enSQLiteInjectedError) != (flowGlobalType)0;
+        TraceEvent("VFSCheckInjectedError")
```

Comment:
Won't this event be too spammy? 
If only used in simulation, we can use SevDebug.

Reply:
It's not spammy.  It is only called in simulation when SQLite returns an error, which is not often.  Also, when this does happen there is usually a flurry of messages as the exception bubbles up, so one more does not make a difference.